File: acmicpc/A02110.py
from sys import stdin as s


def solution(coord: list[int], n: int, c: int) -> None:
    coord.sort()
    low, high, ans = 1, coord[-1] - coord[0], 0
    while low <= high:
        dist, prev, count = (low + high) // 2, coord[0], 1
        for cur in coord[1:]:
            if cur - prev >= dist:
                prev = cur
                count += 1
                if count > c:
                    break
        if count >= c:
            low = dist + 1
            ans = dist
        else:
            high = dist - 1
    print(ans)


def reader():
    n, c = map(int, s.readline().split())
    coord = []
    for _ in range(n):
        coord.append(int(s.readline().strip()))
    solution(coord, n, c)


if __name__ == '__main__':
    reader()

File: acmicpc/A02110.java
package acmicpc;

import java.io.*;
import java.util.*;

public class A02110 {
    public static int solution(int[] coord, int n, int required) {
        Arrays.sort(coord);
        int low = 1, high = coord[n - 1];
        while (low <= high) {
            int dist = low + (high - low) / 2, prev = coord[0], count = 1;
            for (int i = 1; i < n; i++) {
                if (coord[i] - prev >= dist) {
                    prev = coord[i];
                    if (++count >= required) break;
                }
            }
            if (count < required)
                high = dist - 1;
            else
                low = dist + 1;
        }
        return high;
    }

    public void reader() throws NumberFormatException, IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken()), c = Integer.parseInt(st.nextToken());
        int coord[] = new int[n];
        for (int i = 0; i < n; i++)
            coord[i] = Integer.parseInt(br.readLine());
        System.out.println(solution(coord, n, c));
    }

    public static void main(String[] args) throws NumberFormatException, IOException {
        A02110 test = new A02110();
        test.reader();
    }
}
